### ISSUE TITLE
Create service for Search Template API

### ISSUE THREAD

**Issue by olivere** [olivere/elastic#977]
See [the REST API spec](https://github.com/elastic/elasticsearch/blob/6.x/rest-api-spec/src/main/resources/rest-api-spec/api/search_template.json),
the [documentation](https://github.com/elastic/elasticsearch/blob/237650e9c054149fd08213b38a81a3666c1868e5/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/RestSearchTemplateAction.java) and
e.g. the [official .NET client](https://github.com/elastic/elasticsearch-net/blob/67fe8ce6cbaea8ab593dfe77d038fa5b6bb4f903/src/Nest/Search/SearchTemplate/ElasticClient-SearchTemplate.cs).

**Comment by olivere** [comment]
Here's the [code to generate a `SearchTemplateRequest`](https://github.com/elastic/elasticsearch/blob/6.x/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/SearchTemplateRequest.java).

**Comment by prasad-marne** [comment]
i was looking for some example of how to implement searchTemplate POST using elastic library.
Links in this issue are broken. can u link some document where i can search using template.

**Comment by olivere** [comment]
The latest API specification for 6.x can be [found here now](https://github.com/elastic/elasticsearch/blob/6.7/rest-api-spec/src/main/resources/rest-api-spec/api/search_template.json).

I guess the latest implemenation for 6.x [is this](https://github.com/elastic/elasticsearch/blob/6.7/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/SearchTemplateRequestBuilder.java).

**Comment by prasad-marne** [comment]
i would like to implement this feature. I am not sure the best practice so can u guide me how should i go about implementing this service?

**Comment by ghost** [comment]
Hi,

We use olivere/elastic and are looking at using `_search/template`. Therefore, we are interested in this feature. We might also be interested in contributing to this with a PR. Where should we look to find information about how to generate the service?

I found in a comment you made in another issue that you used https://github.com/olivere/elastic/tree/generate-api/generator to generate some code from the api spec.

Thank you.

**Comment by foxen** [comment]
"API Status" marks Search Template as available, this is quite misleading

**Comment by olivere** [comment]
@foxen It's partly implemented. E.g. you should be able to [store, get, and delete a template](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-template.html#pre-registered-templates). What is missing is search and [render](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-template.html#_validating_templates).

**Comment by olivere** [comment]
> Hi,
> 
> We use olivere/elastic and are looking at using `_search/template`. Therefore, we are interested in this feature. We might also be interested in contributing to this with a PR. Where should we look to find information about how to generate the service?
> 
> I found in a comment you made in another issue that you used https://github.com/olivere/elastic/tree/generate-api/generator to generate some code from the api spec.
> 
> Thank you.

@jackdbernier Yes, the code generator was a good starting point. Nowadays, as only few services get added between versions, I will mostly simply take an existing service as a blueprint.

For the implementation part, I _always_ use the Java source as the ultimate truth. The ES documentation is very good and useful, but it's written for humans and is often ambiguous and lacks details you can only find in the source code.
